src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1226 @@
+        assert(ret_n);
+
+        if (strv_isempty(targets))
+                return log_error_errno(SYNTHETIC_ERRNO(ENOENT), "No targets found.");
+
+        STRV_FOREACH(id, targets) {
+                _cleanup_free_ char *bus_path = NULL, *version = NULL;
+
+                r = parse_target(*id, &bus_path, &version);
+                if (r < 0)
+                        return log_oom();
+
+                if (version && !ret_versions)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Unexpected version specifier in target: %s",
+                                               *id);
+
+                r = strv_extend(&bus_paths, strempty(bus_path));
+                if (r < 0)
+                        return r;
+
+                r = strv_extend(&versions, strempty(version));
+                if (r < 0)
+                        return r;
+
+                n++;
+        }
+
+        *ret_n = n;
+        *ret_bus_paths = TAKE_PTR(bus_paths);
+        if (ret_versions)
+                *ret_versions = TAKE_PTR(versions);
+        return 0;
+}
+
+static int log_bus_error(int r, const sd_bus_error *error, const char *target, const char *action) {
+        assert(action);
+
+        if (r == 0) {
+                assert(sd_bus_error_is_set(error));
+                r = sd_bus_error_get_errno(error);
+        }
+
+        if (sd_bus_error_has_name(error, SD_BUS_ERROR_UNKNOWN_OBJECT)) {
+                if (target)
+                        return log_error_errno(r, "Invalid target: %s", target);
+                else
+                        return log_error_errno(r, "Invalid target");
+        }
+
+        if (target)
+                return log_error_errno(r, "Failed to %s for '%s': %s", action, target,
+                                       bus_error_message(error, r));
+        else
+                return log_error_errno(r, "Failed to %s: %s", action, bus_error_message(error, r));
+}
+
+static int list_targets(sd_bus *bus) {
+        _cleanup_(table_unrefp) Table *table = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        _cleanup_strv_free_ char **targets = NULL, **target_paths = NULL;
+        size_t n;
+        int r;
+
+        assert(bus);
+
+        r = ensure_targets(bus, /* argv= */ NULL, &targets);
+        if (r < 0)
+                return r;
+
+        r = parse_targets(targets, &n, &target_paths, /* ret_versions= */ NULL);
+        if (r < 0)
+                return r;
+
+        table = table_new("target", "version", "path");
+        if (!table)
+                return log_oom();
+
+        for (size_t i = 0; i < n; i++) {
+                char *version = NULL, *path = NULL;
+
+                r = sd_bus_call_method(bus, bus_sysupdate_mgr->destination, target_paths[i], SYSUPDATE_TARGET_INTERFACE, "GetVersion", &error, &reply, "");
```

Comment:
You are leaking `reply` here. Please make sure the variable/cleanup declaration is in the right scope